### PR TITLE
Issue #1946: Add a white background (not black) when cropping images …

### PR DESCRIPTION
**Pull request by jenlampton**
Fixes https://github.com/backdrop/backdrop-issues/issues/1946

### CODE REVIEW

File: core/modules/system/image.gd.inc [hunk 1]
```diff
@@ -58,15 +57,31 @@ function image_gd_check_settings() {
  *   The new width of the resized image, in pixels.
  * @param $height
  *   The new height of the resized image, in pixels.
+ * @param $background
+ *   A hexadecimal integer specifying the background color to use for the
+ *   uncovered area of the image after transformation. Examples: "#RGB",
+ *   "#RGBA", "#RRGGBB", "#RRGGBBAA". For images that support transparency, this
+ *   will default to transparent. Otherwise it will be white (#FFFFFF)..
```
Comment (quicksketch) [inline core/modules/system/image.gd.inc]: These examples should exclude the hash (#), as $background will be invalid if it is included.

There's also an extra period at the end.

File: core/modules/system/image.gd.inc [hunk 2]
```diff
@@ -173,15 +188,27 @@ function image_gd_rotate(stdClass $image, $degrees, $background = NULL) {
  *   The width of the cropped area, in pixels.
  * @param $height
  *   The height of the cropped area, in pixels.
+ * @param $background
+ *   A hexadecimal integer specifying the background color to use for the
+ *   uncovered area of the image after transformation. Examples: "#RGB",
+ *   "#RGBA", "#RRGGBB", "#RRGGBBAA". For images that support transparency, this
+ *   will default to transparent. Otherwise it will be white (#FFFFFF)..
```
Comment (quicksketch) [inline core/modules/system/image.gd.inc]: Same as above, # should be excluded and remove the extra period.

File: core/modules/image/image.effects.inc [hunk 3]
```diff
@@ -251,22 +251,16 @@ function image_desaturate_effect(&$image, $data) {
  * @see image_rotate().
  */
 function image_rotate_effect(&$image, $data) {
-  // Set sane default values.
+  // Set sane demn
```
Comment (klonos) [inline core/modules/image/image.effects.inc]: ??? accidental edit?

Comment (jenlampton) [inline core/modules/image/image.effects.inc]: whoops :) good catch!

**Comment by Gormartsen** [comment]
@jenlampton to restart gitlc tests you can close/open this PR

File: core/modules/image/image.effects.inc (hunk 3 above)
Comment (klonos) [inline core/modules/image/image.effects.inc]: :wink:

**Comment by jenlampton** [comment]
testing

**Comment by jenlampton** [comment]
testing

File: core/includes/image.inc [hunk 4]
```diff
@@ -452,5 +451,91 @@ function image_save(stdClass $image, $destination = NULL) {
...
+ *   $image->info['height'] values will be modified by this call.
+ *
+ * @return $background
+ *   A color identifier or FALSE if the allocation failed. If no color is
+ *   provided this will default to transparent For images that support
+ *   transparency, otherwise it will be white.
+ *
+ * @see image_hex2rgba().
+ */
+function image_get_colorindex($background, stdClass &$image) {
```
Comment (quicksketch) [inline core/includes/image.inc]: To match with coding standards, parameters that are by-reference should be first, followed by optional parameters. As `$background` can be optional, it should have a default of NULL and be the second parameter.

`$image` parameter also does not need the `&` by-reference indicator. All objects are passed by reference in PHP 5+.

File: core/includes/image.inc [hunk 5]
```diff
@@ -452,5 +451,91 @@ function image_save(stdClass $image, $destination = NULL) {
...
+  // Save a uppercase version without leading "#" for later processing.
+  $hex_color = drupal_strtoupper(ltrim($hex, '#'));
+
+  // Normalize shorthand versions.
+  // '#FA3' will become '#FFAA33', '#FA37' will become '#FFAA3377'.
+  if (strlen($hex_color) <= 4) {
+    $hex_color = preg_replace('|([0-9A-F])|', '\1\1', $hex_color);
+  }
+
+  // Assure consistency for the color string.
```
Comment (quicksketch) [inline core/includes/image.inc]: `Assure` should be `Ensure`

File: core/modules/simpletest/tests/image.test [hunk 6]
```diff
@@ -323,14 +322,14 @@ class ImageToolkitGdTestCase extends BackdropWebTestCase {
       $operations += array(
         'rotate_5' => array(
           'function' => 'rotate',
-          'arguments' => array(5, 0xFF00FF), // Fuchsia background.
+          'arguments' => array(5, '0xFF00FF'), // Fuchsia background.
```
Comment (quicksketch) [inline core/modules/simpletest/tests/image.test]: This change shouldn't be necessary. If it is, it's an API change. The code may need to adjust to the old argument list if needed.

File: core/modules/system/image.gd.inc [hunk 7]
```diff
@@ -173,15 +158,31 @@ function image_gd_rotate(stdClass $image, $degrees, $background = NULL) {
  *   The width of the cropped area, in pixels.
  * @param $height
  *   The height of the cropped area, in pixels.
+ * @param $background
+ *   A string specifying an RGBA background color to use for the uncovered area
+ *   of the image after manipulation.  E.g. '#RGB', 'RGB', '#RGBA', 'RGBA',
+ *   '#RRGGBB', 'RRGGBB', '#RRGGBBAA', 'RRGGBBAA' or '0xRRGGBBAA'. For images
+ *   that support transparency, this will default to transparent. Otherwise it
+ *   will be white.
```
Comment (quicksketch) [inline core/modules/system/image.gd.inc]: This seems odd that all other functions in this file take hexadecimal parameters but this one takes a string. Is there a reason for this?

File: core/includes/image.inc (hunk 4 above)
Comment (jenlampton) [reply core/includes/image.inc]: updated.

File: core/includes/image.inc (hunk 5 above)
Comment (jenlampton) [reply core/includes/image.inc]: fixed.

File: core/modules/simpletest/tests/image.test (hunk 6 above)
Comment (jenlampton) [inline core/modules/simpletest/tests/image.test]: removed.

File: core/modules/system/image.gd.inc (hunk 7 above)
Comment (jenlampton) [reply core/modules/system/image.gd.inc]: It doesn't look like it, but let me get the tests passing, and see if it still works the same with hex.

**Comment by jenlampton** [comment]
Well, all the tests are passing, but I'm getting an AJAX fatal while running the tests... 

```
An AJAX HTTP error occurred. HTTP Result Code: 200 Debugging information follows. Path: /batch?id=14&op=do_nojs&op=do StatusText: OK ResponseText: {"status":true,"percentage":"100","message":"Processed test 1 of 1 - \u003Cem class=\u0022placeholder\u0022\u003EImage effects\u003C\/em\u003E.\u003Cdiv class=\u0022simpletest-pass\u0022\u003EOverall results: 43 passes, 0 fails, and 0 exceptions\u003C\/div\u003E\u003Cdiv class=\u0022item-list\u0022\u003E\u003Cul\u003E\u003Cli class=\u0022odd first last\u0022\u003E\u003Cdiv class=\u0022simpletest-pass\u0022\u003EImage effects: 43 passes, 0 fails, and 0 exceptions\u003C\/div\u003E\u003C\/li\u003E\u003C\/ul\u003E\u003C\/div\u003E"} backdrop.dev @import url("http://backdrop.dev/core/modules/contextual/css/contextual.css?ocpqyk"); @import url("http://backdrop.dev/core/modules/admin_bar/css/admin_bar.css?ocpqyk"); window.Backdrop = {settings: {"basePath":"\/","pathPrefix":"","drupalCompatibility":true,"admin_bar":{"destination":"destination=batch%3Fid%3D14%26op%3Ddo","hash":"e4ea59f1f5db43fb0bb62641b6746770","basePath":"\/admin_bar","replacements":{".admin-bar-users a":"1"},"margin_top":1},"ajaxPageState":{"css":{"core\/modules\/contextual\/css\/contextual.css":1,"core\/modules\/admin_bar\/css\/admin_bar.css":1},"js":{"core\/misc\/html5.js":1,"core\/misc\/jquery.js":1,"core\/misc\/jquery.once.js":1,"core\/misc\/backdrop.js":1,"core\/modules\/contextual\/js\/contextual.js":1,"core\/modules\/admin_bar\/js\/admin_bar.js":1}}}}; Skip to main content Configure block You are hereHome 
```

there's nothing in the logs. I'll run this test on this site again with core alone and see if the same thing happens.

edit: yes, this is happening with a clean checkout of 1.x. let me reinstall and see if that clears it up.

edit: yes, this is still broken on a fresh install.  Opening a new issue.

File: core/modules/simpletest/tests/image.test (hunk 6 above)
Comment (jenlampton) [reply core/modules/simpletest/tests/image.test]: Interesting, this change does seem to be necessary. Without the quotes the tests fail.

Comment (jenlampton) [reply core/modules/simpletest/tests/image.test]: For some reason PHP is converting `0xFF00FF` into the integer `16711935` and that's what's throwing off the tests. All the functions down the line need to receive a value in the form `#FF00FF`, at least according to their docblocks, so I'm not sure how this ever worked in the first place. Too tired to sort it now, will come back to it tomorrow...

**Comment by jenlampton** [comment]
Rebased and pushed a new PR. All tests still passing for me.

**Comment by quicksketch** [comment]
Replaced by https://github.com/backdrop/backdrop/pull/1698 that fixes the failing tests.

**Pull request closed without merging**